### PR TITLE
Create volumeMount and hostPath for cloud config file

### LINKED ISSUE #576: volumeMount and hostPath are not created for cloud config file
if we specify `cloudProvider: <cloud provider>` in say `kubeadm.conf` and pass that file into `kubeadm init --config=kubeadm.conf` then we add `--cloud-provider` in the api server and controller manager yaml files. Also if the /etc/kubernetes/cloud-config file exists then we add `--cloud-config` command line too.

However we do not create the volumeMount and hostPath for the same. So we need to fix that

### PR DESCRIPTION
**Pull request by dims**
We have a way to specify the cloudProvider in kubeadm.conf. We also
add `--cloud-config /etc/kubernetes/cloud-config` to both the
kubernetes api server and controller manager yaml files if one exists
on the box. However we fail to make that file available to the
process running in the container. We need to make this `cloud-config`
file available to both processes similar to how controller-manager.conf
is passed to controller manager.

**What this PR does / why we need it**:
Fixes https://github.com/kubernetes/kubeadm/issues/576

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: cmd/kubeadm/app/phases/controlplane/volumes.go [hunk 1]
```diff
@@ -70,6 +70,13 @@ func getHostPathVolumesForTheControlPlane(cfg *kubeadmapi.MasterConfiguration) c
 	// Read-only mount for the controller manager kubeconfig file
 	controllerManagerKubeConfigFile := filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.ControllerManagerKubeConfigFileName)
 	mounts.NewHostPathMount(kubeadmconstants.KubeControllerManager, kubeadmconstants.KubeConfigVolumeName, controllerManagerKubeConfigFile, controllerManagerKubeConfigFile, true, &hostPathFileOrCreate)
+	// Read-only mount of the cloud config file if present
+	if cfg.CloudProvider != "" {
+		if _, err := os.Stat(DefaultCloudConfigPath); err == nil {
+			mounts.NewHostPathMount(kubeadmconstants.KubeAPIServer, kubeadmconstants.KubeCloudConfigVolumeName, DefaultCloudConfigPath, DefaultCloudConfigPath, true, &hostPathFileOrCreate)
```
Comment (xiangpengzhao) [inline cmd/kubeadm/app/phases/controlplane/volumes.go]: Should `hostPathFileOrCreate` be `hostPathDirectoryOrCreate`?
or, if we use `hostPathFileOrCreate`, we should mount the cloud config file, not the directory.

**Review by xiangpengzhao — COMMENTED** [review]
just one nit. otherwise LGTM.

File: cmd/kubeadm/app/phases/controlplane/volumes.go (hunk 1 above)
Comment (xiangpengzhao) [reply cmd/kubeadm/app/phases/controlplane/volumes.go]: After taking another looking at the code, seems like `DefaultCloudConfigPath = "/etc/kubernetes/cloud-config"` is a file? Hmm, the file name is a bit confusing...

**Comment by xiangpengzhao** [comment]
Oh, just find that this overlaps with https://github.com/kubernetes/kubernetes/pull/56513/files#diff-d4a6251d485cc68027cd68acd4398701R80. (But this PR is more reasonable than the overlapped part in that one.)

**Comment by dims** [comment]
Ah thanks for the pointer @xiangpengzhao yes, i'd prefer a focused PR like this get in first as it targets a fresh deployment which we need to support better as well. I have to update the docs too

**Comment by dims** [comment]
@luxas would you consider this PR for 1.9 as it is clearly a bug for fresh installs?

File: cmd/kubeadm/app/phases/controlplane/volumes.go (hunk 1 above)
Comment (luxas) [reply cmd/kubeadm/app/phases/controlplane/volumes.go]: should be FileOrCreate yes

Comment (luxas) [reply cmd/kubeadm/app/phases/controlplane/volumes.go]: Is this REALLY needed for the API server? Is there actually anything in the API server that can consume cloud code anymore...? I don't think so

File: cmd/kubeadm/app/constants/constants.go [hunk 2]
```diff
@@ -189,6 +189,9 @@ const (
 	// KubeConfigVolumeName specifies the name for the Volume that is used for injecting the kubeconfig to talk securely to the api server for a control plane component if applicable
 	KubeConfigVolumeName = "kubeconfig"
 
+	// KubeCloudConfigVolumeName specifies the name for the Volume that is used to inject the cloud configuration for kube-apiserver and kube-controller-manager component
+	KubeCloudConfigVolumeName = "kube-cloud-config"
```
Comment (luxas) [inline cmd/kubeadm/app/constants/constants.go]: please define the constant in the same `volumes.go` file as I did in https://github.com/kubernetes/kubernetes/pull/56513

**Review by luxas — APPROVED** [review]
Please make sure the unit test always passes and copy over the change in https://github.com/kubernetes/kubernetes/pull/56513 touching `volumes_test.go`

Small nits only, I approve this bug fix for v1.9

File: cmd/kubeadm/app/constants/constants.go (hunk 2 above)
Comment (dims) [reply cmd/kubeadm/app/constants/constants.go]: Ack on it

**Review by luxas — APPROVED** [review]
/lgtm

**Pull request merged**